### PR TITLE
Track topics used by a Team on Team Broker

### PR DESCRIPTION
**Pull request by hardillb**
part of #4726 
## Description

Tracks all topics published to on the Team Broker.

Adds 1 new HTTP endpoint
- GET `/api/v1/teams/:teamId/broker/topics` returns an array of strings representing the topics the team has published to

Requires team Member or Owner to access list of topics

Needs:
- ~~testing~~
- ~~a new permission adding to access the endpoint~~

Currently caches topics for 1 hour, checks for removing topics every 30seconds, probably should be configurable.

Will benefit from using Redis for this so cache is preserved over restarts and if multiple instances of forge app running.

## Related Issue(s)

#4726 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Review by knolleary — COMMENTED** [review]
Inspired thinking - I like it.

I suggest structuring topic list to have `teamId` has a top level key, rather than just a flat list of all topics.

```
{
   "ff/v1/team1" : {
      "foo": { ttl: 123 }
   },
   "ff/v1/team2" : {
      "bar": { ttl: 123 }
   }
}
```

That will make the `get` operation more scalable as it doesn't have to iterate over the entire cache.

**Comment by hardillb** [comment]
I did it this way to make it easier to include the project nodes changes later (when they will publish without the mount point so will need to include the full topic prefix).

Might need to discuss think about it some more.

**Comment by hardillb** [comment]
Not sure how to write a test for those last 5 lines as that would mean adding a 30 second pause to ensure the interval runs (but even then it won't remove any entries for 30mins)

File: forge/ee/lib/teamBroker/index.js [hunk 1]
```diff
@@ -30,4 +30,50 @@ module.exports.init = function (app) {
...
+    const topicsList = {}
+
+    const cleanInterval = setInterval(() => {
+        const keys = Object.keys(topicsList)
+        const now = Date.now()
+        for (let i = 0; i < keys.length; i++) {
+            if (topicsList[keys[i]].ttl < now) {
+                delete topicsList[keys[i]]
+            }
+        }
```
Comment (knolleary) [inline forge/ee/lib/teamBroker/index.js:50]: This will need updating given the cache structure has changed

**Comment by joepavitt** [comment]
> Currently caches topics for 1 hour, checks for removing topics every 30seconds, probably should be configurable.

Is there an opportunity for a longer term solution here? May be valid that some data only gets published every 24 hours for example.

**Comment by hardillb** [comment]
> Is there an opportunity for a longer term solution here? May be valid that some data only gets published every 24 hours for example.

These are just starting values, but having a longer caching period is likely to not be useful in it's current form as the cache will get cleared every time the forge app restarts (e.g. every time we do a CI deployment....)

There is a comment in the code about using Redis as a better cache, but this is a MUCH bigger solution

**Comment by joepavitt** [comment]
Why is the URL `/api/v1/teams/:teamId/broker/topicList` and not `/api/v1/teams/:teamId/broker/topics`

The latter is more RESTful?

**Comment by knolleary** [comment]
Think about longevity of the cache - and dealing with CI reloading the app.

Some options that come to mind:

1. Introduce redis (more likely valkey). This will also mean we aren't storing state in memory that won't horizontally scale. However it does add the overhead work on adding another component to the architecture. It is inevitable... the question is whether *this* is the feature that pushes us into action on it.
2. Store topic list in existing DB. Don't really want to add this load to the DB in realtime, but could be cached in memory and periodically written to DB in batches (including in the shutdown hook). Doesn't help with scaling (as there is state in memory), but will be persistent across restarts.

In terms of the existing caching TTL values, I'd suggest we default for longer intervals - assuming we (eventually) solve the persistence of state. I'd suggest 25hrs ttl and once a day purging - I don't think the level of usage for in-memory caching will be a problem in the short term for it to be quite generous.

**Comment by hardillb** [comment]
Increasing the cache to 25hours is an easy change, but moving to clearing the cache to once a day, will mean moving that code to the house keeping tasks to get round the restarts.

Will look at that shortly

**Comment by hardillb** [comment]
Stashed the topic cache in the database across restarts

**Comment by hardillb** [comment]
Need to check that this DB cache actually will work on K8s as I think it starts the new instance, before telling the old one to shutdown.

This means the cache will load from the DB before it's populated. Might need to add a delay (and find away to force the `app.settings.get` to re-read from the database.

File: forge/lib/permissions.js [hunk 2]
```diff
@@ -179,7 +179,8 @@ const Permissions = {
     'broker:clients:list': { description: 'List Team Broker clients', role: Roles.Member },
     'broker:clients:create': { description: 'Create Team Broker clients', role: Roles.Owner },
     'broker:clients:edit': { description: 'Edit Team Broker clients', role: Roles.Owner },
-    'broker:clients:delete': { description: 'Delete Team Broker clients', role: Roles.Owner }
+    'broker:clients:delete': { description: 'Delete Team Broker clients', role: Roles.Owner },
+    'broker:topics:list': { description: 'Delete Team Broker clients', role: Roles.Member }
```
Comment (knolleary) [inline forge/lib/permissions.js:183]:
```suggestion
    'broker:topics:list': { description: 'List active Team Broker topics', role: Roles.Member }
```

**Review by knolleary — APPROVED** [review]

**Comment by joepavitt** [comment]
Is this good to merge?

**Pull request merged**